editor/components/post-permalink/index.js

```diff
@@ -14,55 +14,151 @@ import { Dashicon, ClipboardButton, Button } from '@wordpress/components';
+		const regex = /%(?:postname|pagename)%/;
 		this.setState( {
-			showCopyConfirmation: true,
+			samplePermalink: template.replace( regex, postName ),
+			samplePermalinkPrefix: template.split( regex )[ 0 ],
+			samplePermalinkSuffix: template.split( regex )[ 1 ],
+			postName: postName,
 		} );
 	}
 
-	onFinishCopy() {
+	onSavePermalink() {
+		const postName = this.state.editedPostName.replace( /\s+/g, '-' );
+
+		this.props.editPost( {
+			slug: postName,
+			sample_permalink: [ this.props.samplePermalinkData[ 0 ], postName ],
+		} );
+
 		this.setState( {
-			showCopyConfirmation: false,
+			editingPermalink: false,
+			editedPostName: postName,
 		} );
 	}
 
 	render() {
-		const { isNew, link } = this.props;
-		if ( isNew || ! link ) {
+		const { isNew, previewLink, permalinkStructure } = this.props;
+		const {
+			iconClass,
+			samplePermalink,
+			samplePermalinkPrefix,
+			samplePermalinkSuffix,
+			postName,
+			editingPermalink,
+		} = this.state;
+
+		if ( isNew || ! previewLink ) {
 			return null;
 		}
 
 		return (
 			<div className="editor-post-permalink">
-				<Dashicon icon="admin-links" />
-				<span className="editor-post-permalink__label">{ __( 'Permalink:' ) }</span>
-				<Button className="editor-post-permalink__link" href={ link } target="_blank">
-					{ decodeURI( link ) }
-				</Button>
 				<ClipboardButton
-					className="button"
-					text={ link }
-					onCopy={ this.onCopy }
-					onFinishCopy={ this.onFinishCopy }
+					className="editor-post-permalink__copy"
+					text={ samplePermalink }
+					title={ __( 'Copy the permalink to your clipboard.' ) }
+					onCopy={ () => this.setState( { iconClass: 'copied' } ) }
 				>
-					{ this.state.showCopyConfirmation ? __( 'Copied!' ) : __( 'Copy' ) }
+					<Dashicon icon="admin-links" className={ iconClass } />
 				</ClipboardButton>
+
+				<span className="editor-post-permalink__label">{ __( 'Permalink:' ) }</span>
+
+				{ // Show the sample permalink when it isn't being editing.
+					! editingPermalink &&
+					<Button className="editor-post-permalink__link" href={ previewLink } target="_blank">
+						{ decodeURI( samplePermalink ) }
+					</Button>
+				}
+
+				{ // When we're editing the permalink, show the edit form.
+					editingPermalink &&
+					<form
+						className="editor-post-permalink__edit-form"
+						onSubmit={ this.onSavePermalink }>
+						<span className="editor-post-permalink__prefix">
+							{ samplePermalinkPrefix }
+						</span>
+						<input
+							type="text"
+							className="editor-post-permalink__edit"
+							defaultValue={ postName }
+							onChange={ ( e ) => this.setState( { editedPostName: e.target.value } ) }
+							required
+						/>
+						<span className="editor-post-permalink__suffix">
+							{ samplePermalinkSuffix }
+						</span>
+						<Button
+							className="editor-post-permalink__save button"
+							onClick={ this.onSavePermalink }
+						>
+							{ __( 'Ok' ) }
```

Comment:
Minor / Question: At least one prior instance of going out of the way to capitalize "OK":

https://github.com/WordPress/WordPress/blob/0895f04705442bba10f9f4afd3b4dce36aa8f6a7/wp-includes/class-wp-editor.php#L1063

Reply:
👍🏻